ibis/expr/format.py

```diff
@@ -1,283 +1,598 @@
+    if depth and depth < len(pieces):
+        return fmt_truncated(pieces, depth=depth)
+    return "\n\n".join(pieces)
+
+
+def fmt_truncated(
+    pieces: Iterable[str],
+    *,
+    depth: int,
+    sep: str = "\n\n",
+    ellipsis: str = util.VERTICAL_ELLIPSIS,
+) -> str:
+    if depth == 1:
+        return pieces[-1]
+
+    first_n = depth // 2
+    last_m = depth - first_n
+    return sep.join([*pieces[:first_n], ellipsis, *pieces[-last_m:]])
+
+
+def selection_maxlen(expressions: Iterable[ir.ValueExpr]) -> int:
+    """Compute the length of the longest name of input expressions.
 
     Parameters
     ----------
-    get_text_repr: bool
-         Defaults to ``False``. Determines whether or not the memoization
-         should use proper alias names. Using the same alias names for
-         equivalent expressions is more optimal for memoization / recursion
-         but does not accurately display aliases in the representation
+    expressions
+        Expressions whose name to compute the maximum length of
+
+    Returns
+    -------
+    int
+        Max length
     """
+    try:
+        return max(
+            len(name)
+            for expr in expressions
+            if (name := expr._safe_name) is not None
+        )
+    except ValueError:
+        return 0
+
+
+@functools.singledispatch
+def fmt_generic(op: ops.Node, *, aliases: Aliases, **_: Any) -> str:
+    """Fallback formatting implementation."""
+    raw_parts = fmt_fields(
+        op,
+        dict.fromkeys(op.__signature__.parameters.keys(), fmt_value),
+        aliases=aliases,
+    )
+    return f"{op.__class__.__name__}\n{raw_parts}"
+
+
+@fmt_generic.register
+def _(op: ops.TableNode, **kwargs: Any) -> str:
+    return fmt_table_op(op, **kwargs)
+
+
+@fmt_generic.register
+def _(op: ops.ValueOp, *, name: str, aliases: Aliases, **_: Any) -> str:
+    value = fmt_value(op, aliases=aliases)
+    prefix = f"{name}: " if name is not None else ""
+    return f"{prefix}{value}{type_info(op.to_expr().type())}"
+
 
-    def __init__(self, get_text_repr: bool = False):
-        self.formatted: Dict[str, Any] = {}
-        self.aliases: Dict[str, str] = {}
-        self.ops: Dict[str, ops.Node] = {}
-        self.counts: Dict[str, int] = defaultdict(int)
-        self._repr_memo: Dict[ir.Expr, str] = {}
-        self.subexprs: Dict[ops.Node, str] = {}
-        self.visit_memo: Set[ops.Node] = set()
-        self.get_text_repr = get_text_repr
+@fmt_generic.register
+def _(op: ops.SortKey, *, aliases: Aliases, **_: Any) -> str:
+    return fmt_value(op, aliases=aliases)
 
-    def __contains__(self, obj):
-        return self._key(obj) in self.formatted
 
-    def _key(self, expr) -> str:
-        memo = self._repr_memo
-        try:
-            result = memo[expr]
-        except KeyError:
-            result = memo[expr] = self._format(expr)
-        return result
+@functools.singledispatch
+def fmt_table_op(op: ops.TableNode, **_: Any) -> str:
```

Comment:
Why introducing more dispatched functions instead of using a single one: `fmt_generic`

Reply:
`fmt_generic` is only for handling the root expression